src/coord/src/catalog.rs

```diff
@@ -382,25 +382,92 @@ impl CatalogEntry {
+
+        struct TypeNormalizer {
+            err: Option<Error>,
+        }
+
+        impl<'ast> VisitMut<'ast> for TypeNormalizer {
+            fn visit_data_type_mut(&mut self, data_type: &'ast mut DataType) {
+                if let DataType::Other { name, .. } = data_type {
+                    if name.0.len() == 1 {
+                        *name = ObjectName(vec![Ident::new(PG_CATALOG_SCHEMA), name.0.remove(0)]);
+                    }
+                }
+            }
+        }
+
         let mut storage = catalog.storage();
         let items = storage.load_items()?;
         let tx = storage.transaction()?;
+
         for (id, name, def) in items {
-            let item = match catalog.deserialize_item(def) {
-                Ok(item) => item,
-                Err(e) => {
-                    return Err(Error::new(ErrorKind::Corruption {
-                        detail: format!("failed to deserialize item {} ({}): {}", id, name, e),
-                    }))
+            let SerializedCatalogItem::V1 {
+                create_sql,
+                eval_env,
+            } = serde_json::from_slice(&def)?;
+
+            let mut stmt = sql::parse::parse(&create_sql)?.into_element();
+            match &mut stmt {
+                Statement::CreateTable(CreateTableStatement {
+                    name: _,
+                    columns,
+                    constraints: _,
+                    with_options: _,
+                    if_not_exists: _,
+                }) => {
+                    let mut normalizer = TypeNormalizer { err: None };
+                    for c in columns {
+                        normalizer.visit_column_def_mut(c);
+                    }
+                    if let Some(err) = normalizer.err {
+                        return Err(err.into());
+                    }
                 }
+
+                Statement::CreateView(CreateViewStatement {
+                    name: _,
+                    columns: _,
+                    query,
+                    temporary: _,
+                    materialized: _,
+                    if_exists: _,
+                    with_options: _,
+                }) => {
+                    let mut normalizer = TypeNormalizer { err: None };
+                    normalizer.visit_query_mut(query);
+                    if let Some(err) = normalizer.err {
+                        return Err(err.into());
+                    }
+                }
+                // Only tables and views contain references to types; other
+                // objects don't need to be rewritten.
+                _ => continue,
+            }
+
+            let serialized_item = SerializedCatalogItem::V1 {
+                create_sql: stmt.to_ast_string(),
+                eval_env,
             };
-            let serialized_item = catalog.serialize_item(&item);
+
+            let serialized_item =
+                serde_json::to_vec(&serialized_item).expect("catalog serialization cannot fail");
```

Comment:
These have gotten substantial enough that I'd advocate for moving them to a `catalog/migrations.rs` module that just exports the const!